google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,371 @@
+          if (isSplit) {
+            validate(
+                !chunk.getCommitRow(), "NewRow can't commit when valueSize indicates more data");
+            validate(
+                !chunk.getValue().isEmpty(),
+                "NewRow must have data when valueSize promises more data in the next chunk");
+
+            expectedCellSize = chunk.getValueSize();
+            remainingCellBytes = expectedCellSize - chunk.getValue().size();
+          } else {
+            expectedCellSize = chunk.getValue().size();
+            remainingCellBytes = 0;
+          }
+
+          // Start building cell
+          adapter.startCell(familyName, qualifier, timestamp, labels, expectedCellSize);
+          adapter.cellValue(chunk.getValue());
+
+          // Transitions
+          if (isSplit) {
+            return CellInProgress;
+          }
+          adapter.finishCell();
+
+          if (!chunk.getCommitRow()) {
+            return NewCell;
+          }
+
+          return handleCommit();
+        }
+      };
+
+  /** A state that represents a cell's value continuation. */
+  private final State CellInProgress =
+      new State() {
+        @Override
+        State handleChunk(CellChunk chunk) {
+          if (chunk.getResetRow()) {
+            return handleResetChunk(chunk);
+          }
+
+          boolean isLast = chunk.getValueSize() == 0;
+
+          validate(!chunk.hasFamilyName(), "CellInProgress can't have a family");
+          validate(!chunk.hasQualifier(), "CellInProgress can't have a qualifier");
+          validate(chunk.getTimestampMicros() == 0, "CellInProgress can't have a timestamp");
+          validate(chunk.getLabelsCount() == 0, "CellInProgress can't have labels");
+          if (isLast) {
+            long missingBytes = remainingCellBytes - chunk.getValue().size();
+            validate(
+                missingBytes == 0,
+                "CellInProgress is last, but is missing " + missingBytes + " bytes");
+          } else {
+            validate(
+                expectedCellSize == chunk.getValueSize(),
+                "CellInProgress valueSizes should be identical for nonterminal chunks");
+            validate(!chunk.getCommitRow(), "CellInProgress can't commit with a nonterminal chunk");
+          }
+          remainingCellBytes -= chunk.getValue().size();
+          adapter.cellValue(chunk.getValue());
+
+          // Transitions
+          if (!isLast) {
+            return CellInProgress;
+          }
+          adapter.finishCell();
+
+          if (!chunk.getCommitRow()) {
+            return NewCell;
+          }
+
+          return handleCommit();
+        }
+      };
+
+  private static final State CompleteRow =
+      new State() {
+        @Override
+        State handleChunk(CellChunk chunk) {
+          throw new IllegalStateException("Skipping completed row");
+        }
+      };
+
+  // Helpers ------------------------
+  private State handleResetChunk(CellChunk cellChunk) {
```

Comment:
Maybe we should call it handleResetRow()?
It is really discarding what we have read so far for the row and start over. Should document it.

Reply:
I prefer to keep it as is. It's an event handler for special chunks.